lib/tower-http-digest/src/lib.rs

```diff
@@ -0,0 +1,186 @@
+use axum_core::{body::Body, RequestExt};
+use bytes::{BufMut, BytesMut};
+use http::{HeaderName, HeaderValue, Request};
+use http_body::Body as HttpBody;
+use pin_project_lite::pin_project;
+use sha2::{Digest, Sha256, Sha512};
+use std::{
+    error::Error as StdError,
+    future::Future,
+    pin::Pin,
+    str::FromStr,
+    task::{self, ready, Poll},
+};
+use strum::{AsRefStr, EnumString};
+use tower_layer::Layer;
+use tower_service::Service;
+
+type BoxError = Box<dyn StdError + Send + Sync>;
+
+static DIGEST_HEADER_NAME: HeaderName = HeaderName::from_static("digest");
+
+#[derive(AsRefStr, Clone, Copy, Default, EnumString)]
+#[non_exhaustive]
+pub enum Algorithm {
+    #[default]
+    #[strum(ascii_case_insensitive, serialize = "sha-256")]
+    Sha256,
+
+    #[strum(ascii_case_insensitive, serialize = "sha-512")]
+    Sha512,
+}
+
+impl Algorithm {
+    pub fn digest(&self, data: impl AsRef<[u8]>) -> Vec<u8> {
+        match self {
+            Self::Sha256 => Sha256::digest(data).to_vec(),
+            Self::Sha512 => Sha512::digest(data).to_vec(),
+        }
+    }
+}
+
+pin_project! {
+    #[project = DigestFutureProj]
+    pub enum DigestFuture<S, F> {
+        BuildingDigest {
+            service: S,
+
+            algorithm: Algorithm,
+            parts: Option<http::request::Parts>,
+
+            #[pin]
+            body: Body,
+            body_accumulator: Option<BytesMut>,
+        },
+        PollServiceFuture {
+            #[pin]
+            future: F,
+        },
+    }
+}
+
+impl<S> Future for DigestFuture<S, S::Future>
+where
+    S: Service<Request<Body>>,
+    S::Error: Into<BoxError>,
+{
+    type Output = Result<S::Response, BoxError>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut task::Context<'_>) -> Poll<Self::Output> {
+        loop {
+            match self.as_mut().project() {
+                DigestFutureProj::BuildingDigest {
+                    service,
+                    algorithm,
+                    parts,
+                    mut body,
+                    body_accumulator,
+                } => {
+                    while let Some(frame) = ready!(body.as_mut().poll_frame(cx))
+                        .transpose()
+                        .map_err(BoxError::from)?
+                    {
+                        if let Ok(data) = frame.into_data() {
+                            let accumulator = body_accumulator
+                                .as_mut()
+                                .expect("[Bug] Missing accumulator");
+
+                            accumulator.put(data);
+                        }
+                    }
```

Comment:
Interesting. I've thought of a similar approach for this middleware, but since I was focused on rebuilding the digest header itself, this seemed rather infeasible to do without buffering the entire body.  

But erroring out inside the `Body` implementation seems like an interesting idea..

Reply:
Actually, I'll probably just transition to this method. It works, it's cleaner, and more generic.

Reply:
Okay, I ported it over to working with HTTP bodies